python/ray/serve/tests/test_deploy.py

```diff
@@ -1193,6 +1195,73 @@ async def __call__(self):
     assert requests.get(url).text == "2"
 
 
+class TestDeployGroup:
+    @serve.deployment
+    def f(request=None):
+        return "f reached"
+
+    @serve.deployment
+    def g(request=None):
+        return "g reached"
+
+    @serve.deployment
+    class C:
+        async def __call__(self):
+            return "C reached"
+
+    @serve.deployment
+    class D:
+        async def __call__(self):
+            return "D reached"
+
+    @serve.deployment
+    class E:
+        async def __init__(self):
+            self.F_handle = serve.get_deployment("F").get_handle()
+
+        async def __call__(self):
+            return await self.F_handle.request_output.remote()
+
+        async def request_output(self):
+            return await self.F_handle.get_output.remote()
+
+        async def get_output(self):
+            return "E reached"
+
+    @serve.deployment
+    class F:
+        async def __init__(self):
+            self.E_handle = serve.get_deployment("E").get_handle()
+
+        async def __call__(self):
+            return await self.E_handle.request_output.remote()
+
+        async def request_output(self):
+            return await self.E_handle.get_output.remote()
+
+        async def get_output(self):
+            return "F reached"
+
+    deployments = [f, g, C, D]
+    responses = ["f reached", "g reached", "C reached", "D reached"]
+
+    def test_basic_deploy_group(self, serve_instance):
+
+        deploy_group(self.deployments)
+
+        for deployment, response in zip(self.deployments, self.responses):
+            assert ray.get(deployment.get_handle().remote()) == response
+
+    def test_mutual_handles(self, serve_instance):
+        deployments = [self.E, self.F]
+        responses = ["E reached", "F reached"]
+
+        deploy_group(deployments)
+
+        for deployment, response in zip(deployments, responses):
+            assert ray.get(deployment.get_handle().remote()) == response
+
```

Comment:
do we have implementation and test for the atomic behavior ? Currently if a new deploy() failed we will automatically roll back to latest known good target state.

Reply:
Discussed offline– we're planning on pulling the roll back behavior out of the Serve controller and moving it to a separate layer of abstraction. For this change, we can proceed without worrying about how roll back works with `deploy_group` as long as the roll back functionality is removed soon after.

Reply:
yep makes sense, i assume it's out of scope for current planned items for operationalizing serve ?

Reply:
yep

Reply:
see https://github.com/ray-project/ray/pull/22121/files